File: crates/harness-core/src/subagent.rs
//! SubAgent event channel — typed "what is the subagent doing right now" stream.
//!
//! Companion to [`crate::progress`] (per-tool stdout/stderr) and
//! [`crate::plan`] (full plan snapshots). The wire model is the
//! same: a tool installs an [`mpsc::UnboundedSender<SubAgentFrame>`]
//! in a `tokio::task_local` before invoking the subagent body via
//! [`with_subagent`], the subagent emits via [`emit`], and the agent
//! loop drains the receiver alongside `tool.invoke` and forwards
//! each frame as `AgentEvent::SubAgentEvent`.
//!
//! The point of *streaming* (versus "subagent runs and returns a
//! single string when done") is product: users want to **see** the
//! subagent's reasoning, tool calls, and intermediate output while
//! it works — both in an inline collapsible card in the main message
//! stream and on a side panel listing every running subagent. Both
//! consumers read the same event stream.
//!
//! Outside an agent invocation the channel is absent — emits become
//! no-ops, which keeps subagent unit tests trivial (no agent loop
//! required to run a subagent in isolation).

use serde::{Deserialize, Serialize};
use tokio::sync::mpsc;

/// One frame in a subagent's execution stream. The `subagent_id` is
/// a per-invocation uuid (assigned by the harness when the
/// subagent's `Tool::invoke` starts) so the UI can correlate frames
/// from concurrent subagents on the same conversation.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SubAgentFrame {
    /// Unique id for this subagent invocation. Stable across the
    /// whole `Started…Done|Error` sequence; opaque to the renderer.
    pub subagent_id: String,
    /// Subagent registry name (e.g. `"claude_code"`, `"review"`).
    /// Stable across runs of the same subagent kind, so the UI can
    /// pick an icon / label per kind.
    pub subagent_name: String,
    /// The actual event payload.
    pub event: SubAgentEvent,
}

/// A point-in-time event in a subagent's run. Mirrors the shape of
/// the main `AgentEvent` enum but cut down to what a subagent can
/// reasonably surface; the serialiser uses `kind` as the
/// discriminator so UIs can match on it directly.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum SubAgentEvent {
    /// The subagent has begun work. Emitted exactly once per
    /// invocation, immediately after the harness installs the
    /// channel. `model` is informational (which LLM is doing the
    /// thinking, when known — empty for SDK-sidecar subagents whose
    /// model is determined by the SDK itself).
    Started {
        task: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        model: Option<String>,
    },
    /// A chunk of assistant-visible text from the subagent's LLM.
    /// Renderers concatenate consecutive `Delta`s.
    Delta { text: String },
    /// The subagent invoked one of its own tools.
    ToolStart {
        name: String,
        arguments: serde_json::Value,
    },
    /// The subagent's tool returned. `output` mirrors what the inner
    /// tool reported — tool errors are surfaced as text here too,
    /// matching the main loop's behaviour.
    ToolEnd { name: String, output: String },
    /// SDK-style status update — a one-liner the subagent's wrapper
    /// chose to surface (e.g. "Searching files…" from
    /// `@anthropic-ai/claude-agent-sdk`'s structured events). The
    /// renderer can show these as muted lines in the inline card.
    Status { message: String },
    /// The subagent finished cleanly. `final_message` is the same
    /// string returned to the caller via `Tool::invoke`. The matching
    /// `ToolEnd` in the *outer* stream will carry the same text, so
    /// transports may dedupe if they want.
    Done { final_message: String },
    /// The subagent failed. `final_message` returned to the caller is
    /// `format!("subagent error: {message}")` so the outer model can
    /// adapt — same convention as tool errors elsewhere.
    Error { message: String },
}

tokio::task_local! {
    /// Per-invocation subagent sender, scoped via [`with_subagent`].
    static SUBAGENT_TX: mpsc::UnboundedSender<SubAgentFrame>;
}

/// Publish a frame. No-op when no listener is installed (i.e. the
/// subagent was invoked outside an agent loop — the unit-test path).
pub fn emit(frame: SubAgentFrame) {
    let _ = SUBAGENT_TX.try_with(|tx| {
        let _ = tx.send(frame);
    });
}

/// Whether a subagent sender is installed for the current task. Used
/// by subagent implementations to decide whether to bother building
/// frames (frames carry serde_json::Value args and can be costly).
pub fn is_active() -> bool {
    SUBAGENT_TX.try_with(|_| ()).is_ok()
}

/// Run `fut` with `tx` installed as the active subagent sender.
/// Used by the agent loop to scope a sender to a single subagent
/// tool invocation, mirroring [`crate::plan::with_plan`]. The sender
/// goes out of scope when `fut` completes, so subsequent invocations
/// need their own channel.
pub async fn with_subagent<F, R>(tx: mpsc::UnboundedSender<SubAgentFrame>, fut: F) -> R
where
    F: std::future::Future<Output = R>,
{
    SUBAGENT_TX.scope(tx, fut).await
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn emit_inside_with_subagent_reaches_receiver() {
        let (tx, mut rx) = mpsc::unbounded_channel();
        with_subagent(tx, async {
            assert!(is_active());
            emit(SubAgentFrame {
                subagent_id: "sub-1".into(),
                subagent_name: "review".into(),
                event: SubAgentEvent::Started {
                    task: "verify the kanban renders".into(),
                    model: Some("claude-sonnet-4-6".into()),
                },
            });
        })
        .await;
        let frame = rx.try_recv().unwrap();
        assert_eq!(frame.subagent_id, "sub-1");
        match frame.event {
            SubAgentEvent::Started { task, model } => {
                assert_eq!(task, "verify the kanban renders");
                assert_eq!(model.as_deref(), Some("claude-sonnet-4-6"));
            }
            other => panic!("unexpected event: {other:?}"),
        }
    }

    #[tokio::test]
    async fn emit_outside_scope_is_noop() {
        assert!(!is_active());
        emit(SubAgentFrame {
            subagent_id: "x".into(),
            subagent_name: "x".into(),
            event: SubAgentEvent::Status { message: "x".into() },
        });
        // The point: this didn't panic.
    }

    #[test]
    fn event_serialises_with_kind_discriminator() {
        let f = SubAgentFrame {
            subagent_id: "sub-1".into(),
            subagent_name: "claude_code".into(),
            event: SubAgentEvent::ToolStart {
                name: "fs.read".into(),
                arguments: serde_json::json!({ "path": "/etc/hosts" }),
            },
        };
        let json = serde_json::to_string(&f).unwrap();
        assert!(json.contains("\"kind\":\"tool_start\""), "got: {json}");
        assert!(json.contains("\"subagent_id\":\"sub-1\""), "got: {json}");
        assert!(json.contains("\"name\":\"fs.read\""), "got: {json}");
    }
}

File: packages/subagents/src/subagent.ts
// SubAgent surface + event channel. Ported from harness-subagents/src/lib.rs
// (the trait + registry + input/output/artifact types) and
// harness-core/src/subagent.rs (the SubAgentFrame / SubAgentEvent stream types
// + the task-local `with_subagent` channel) and harness-core/src/workspace.rs
// (the `active_workspace` session-workspace seam).
//
// Why the channel + workspace helpers live HERE and not in @jarvis/core: the
// Rust crate *re-exports* them from core, but the Node @jarvis/core port has
// not yet landed `subagent.rs` / `workspace.rs`. To keep this package
// self-contained we define them locally (mirroring the Rust wire shapes
// exactly) and re-export so a later @jarvis/core port can adopt them verbatim.
//
// The wire model mirrors `progress.ts` in @jarvis/core: a tool installs a
// sink in an AsyncLocalStorage (Rust's `tokio::task_local`) before invoking
// the subagent body via `withSubagent`, the subagent emits via `emitSubagent`,
// and the agent loop drains the listener and forwards each frame as
// `AgentEvent::SubAgentEvent`. Outside a scope, emits are no-ops.
import { AsyncLocalStorage } from "node:async_hooks";
import type { JsonValue } from "@jarvis/core";
import type { Usage } from "@jarvis/core";

// ---------------------------------------------------------------------------
// SubAgent input / output / artifacts (harness-subagents/src/lib.rs)
// ---------------------------------------------------------------------------

/** What the main agent hands a subagent when it invokes it. */
export interface SubAgentInput {
  /** Free-form natural-language task. The subagent interprets it. */
  task: string;
  /**
   * Sandbox root the subagent may read / write within. Mirrors the main
   * agent's workspace root. Subagents should refuse paths outside it.
   */
  workspace_root: string;
  /** Optional structured context — verification_plan, run id, etc. */
  context?: JsonValue;
  /**
   * Recursion guard: every nested subagent invocation appends its id. v1.0
   * forbids depth >= 1 from registering further `subagent.*` tools, but the
   * chain travels with frames so the UI can draw nesting later. Absent when
   * empty (matches serde `skip_serializing_if = "Vec::is_empty"`).
   */
  caller_chain?: string[];
}

/**
 * A typed product a subagent can attach to its output. Discriminated on
 * `kind` (serde `tag = "kind", rename_all = "snake_case"`). Renderers treat
 * unknown variants as "ignore" for forward-compat.
 */
export type Artifact =
  | { kind: "files_changed"; paths: string[] }
  | { kind: "review_verdict"; pass: boolean; commentary: string; evidence: string[] }
  | { kind: "doc_summary"; summary: string; quotes: string[] };

/**
 * What a subagent returns when it finishes. `message` is the textual output
 * the main agent sees (matching the `Done.final_message` frame); `artifacts`
 * are optional structured products. Absent when empty (serde skip).
 */
export interface SubAgentOutput {
  message: string;
  artifacts?: Artifact[];
}

/**
 * A delegated agent the main Agent may invoke through a `subagent.<name>`
 * tool. The Node analogue of Rust's `#[async_trait] trait SubAgent`.
 */
export interface SubAgent {
  /** Registry key. Becomes the `subagent.<name>` tool name. */
  readonly name: string;
  /** One-line description shown to the main agent in the tool spec. */
  readonly description: string;
  /** JSON schema for the tool input (object-shaped). */
  inputSchema(): JsonValue;
  /**
   * Whether the wrapping `subagent.<name>` tool is approval-gated. Subagents
   * that mutate the workspace (claude_code / codex) must return true;
   * read-only subagents (read_doc / review) return false. Default false.
   */
  requiresApproval?(): boolean;
  /**
   * Run the subagent. Implementations emit `started` early, stream
   * `delta` / `tool_start` / `tool_end` / `status` frames, and exactly one
   * `done` or `error` at the end. The harness wraps the call in a
   * `withSubagent(...)` scope so `emitSubagent` reaches the main loop.
   */
  invoke(input: SubAgentInput): Promise<SubAgentOutput>;
}

/** Read a subagent's approval gate, applying the trait's `false` default. */
export function subAgentRequiresApproval(sub: SubAgent): boolean {
  return sub.requiresApproval ? sub.requiresApproval() : false;
}

// ---------------------------------------------------------------------------
// Registry (harness-subagents/src/lib.rs::SubAgentRegistry)
// ---------------------------------------------------------------------------

/**
 * Registry of available subagents. Held by the composition root; the
 * `subagentTools` helper folds each subagent into a `Tool[]`.
 *
 * Last-write-wins on `name`: a later `register` with the same name replaces
 * the prior entry, so the composition root is the single source of truth
 * (collisions are intentional overrides, not silent bugs).
 */
export class SubAgentRegistry {
  // Insertion order preserved; `register` removes any prior same-named entry
  // before pushing, so iteration order reflects last registration.
  #entries: SubAgent[] = [];

  register(sub: SubAgent): this {
    this.#entries = this.#entries.filter((s) => s.name !== sub.name);
    this.#entries.push(sub);
    return this;
  }

  /** All registered subagents, ordered by (latest) insertion. */
  iter(): readonly SubAgent[] {
    return this.#entries;
  }

  /** Look up by name; the most-recently registered entry wins. */
  get(name: string): SubAgent | undefined {
    for (let i = this.#entries.length - 1; i >= 0; i--) {
      const s = this.#entries[i];
      if (s && s.name === name) return s;
    }
    return undefined;
  }

  isEmpty(): boolean {
    return this.#entries.length === 0;
  }

  get length(): number {
    return this.#entries.length;
  }
}

// ---------------------------------------------------------------------------
// Event channel (harness-core/src/subagent.rs)
// ---------------------------------------------------------------------------

/**
 * One frame in a subagent's execution stream. `subagent_id` is a
 * per-invocation uuid (assigned when `invoke` starts) so the UI can correlate
 * frames from concurrent subagents on the same conversation.
 */
export interface SubAgentFrame {
  subagent_id: string;
  subagent_name: string;
  event: SubAgentEvent;
}

/**
 * A point-in-time event in a subagent's run. Discriminated on `kind` (serde
 * `tag = "kind", rename_all = "snake_case"`). `model` on `started` is absent
 * when unknown (serde `skip_serializing_if = "Option::is_none"`); `usage`
 * fields are flattened onto the `usage` variant.
 */
export type SubAgentEvent =
  | { kind: "started"; task: string; model?: string }
  | { kind: "delta"; text: string }
  | { kind: "tool_start"; name: string; arguments: JsonValue }
  | { kind: "tool_end"; name: string; output: string }
  | ({ kind: "usage"; model: string } & Usage)
  | { kind: "status"; message: string }
  | { kind: "done"; final_message: string }
  | { kind: "error"; message: string };

/** A sink that receives emitted subagent frames. */
export type SubAgentSink = (frame: SubAgentFrame) => void;

const channel = new AsyncLocalStorage<SubAgentSink>();

/** Publish a frame. No-op outside a `withSubagent` scope (unit-test path). */
export function emitSubagent(frame: SubAgentFrame): void {
  channel.getStore()?.(frame);
}

/** Whether a subagent sink is installed for the current async context. */
export function subagentActive(): boolean {
  return channel.getStore() !== undefined;
}

/**
 * The currently-installed sink, if any. Used by the parallel-batch tool so
 * each spawned child can re-enter a fresh `withSubagent` scope and forward
 * frames into the same outer transport. `undefined` outside a scope.
 */
export function activeSender(): SubAgentSink | undefined {
  return channel.getStore();
}

/** Run `fn` with `sink` installed as the active subagent sink. */
export function withSubagent<T>(sink: SubAgentSink, fn: () => Promise<T>): Promise<T> {
  return channel.run(sink, fn);
}

// ---------------------------------------------------------------------------
// Session-workspace seam (harness-core/src/workspace.rs)
// ---------------------------------------------------------------------------
//
// The Rust `active_workspace_or` reads a task-local pinned by the agent loop's
// `with_session_workspace`. The Node @jarvis/core port has not landed this
// yet, so we provide a local AsyncLocalStorage seam. The batch + tool-adapter
// fall back to a construction-time `default_root` when no scope is active —
// identical to Rust's `active_workspace_or(default)`.

const workspace = new AsyncLocalStorage<string>();

/** The active session workspace, or `undefined` outside a scope. */
export function activeWorkspace(): string | undefined {
  return workspace.getStore();
}

/** The active session workspace, or `fallback` when none is pinned. */
export function activeWorkspaceOr(fallback: string): string {
  return workspace.getStore() ?? fallback;
}

/** Run `fn` with `root` pinned as the active session workspace. */
export function withSessionWorkspace<T>(root: string | undefined, fn: () => Promise<T>): Promise<T> {
  if (root === undefined) return fn();
  return workspace.run(root, fn);
}
